x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexer.java

```diff
@@ -226,6 +293,72 @@ protected boolean handleCircuitBreakingException(Exception e) {
         return true;
     }
 
+    private void getChangedBuckets(DataFrameTransformCheckpoint oldCheckpoint, DataFrameTransformCheckpoint newCheckpoint,
+            ActionListener<Map<String, List<String>>> listener) {
+        SearchRequest searchRequest = new SearchRequest(getConfig().getSource().getIndex());
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder();
+
+        // we do not need the sub-aggs
+        CompositeAggregationBuilder changesAgg = pivot.buildChangedBucketsAggregation(pageSize);
+        sourceBuilder.aggregation(changesAgg);
+        sourceBuilder.size(0);
+
+        QueryBuilder pivotQueryBuilder = getConfig().getSource().getQueryConfig().getQuery();
+
+        DataFrameTransformConfig config = getConfig();
+        if (config.getSyncConfig() != null) {
+            BoolQueryBuilder filteredQuery = new BoolQueryBuilder().
+                    filter(pivotQueryBuilder).
+                    filter(config.getSyncConfig().getRangeQuery(oldCheckpoint, newCheckpoint));
+
+            logger.trace("Gathering changes using query {}", filteredQuery);
+            sourceBuilder.query(filteredQuery);
+        } else {
+            logger.trace("No sync configured");
+            listener.onResponse(null);
+            return;
+        }
+
+        searchRequest.source(sourceBuilder);
+        searchRequest.allowPartialSearchResults(false);
+
+        // pre-fill the hashmap with the expected keys
+        // note that changed bucket might be empty due to the query
+        HashMap<String, List<String>> keys = new HashMap<>();
+        for(Entry<String, SingleGroupSource> entry: config.getPivotConfig().getGroupConfig().getGroups().entrySet()) {
+            keys.put(entry.getKey(), new ArrayList<>());
+        }
+
+        collectChangedBuckets(searchRequest, changesAgg, keys, ActionListener.wrap(listener::onResponse, e -> {
+            // fall back if bucket collection failed
+            logger.error("Failed to retrieve changed buckets, fall back to complete retrieval", e);
+            listener.onResponse(null);
+        }));
+    }
+
+    void collectChangedBuckets(SearchRequest searchRequest, CompositeAggregationBuilder changesAgg, Map<String, List<String>> keys,
+            ActionListener<Map<String, List<String>>> finalListener) {
+
+        // re-using the existing search hook
+        doNextSearch(searchRequest, ActionListener.wrap(searchResponse -> {
+            final CompositeAggregation agg = searchResponse.getAggregations().get(COMPOSITE_AGGREGATION_NAME);
+
+            agg.getBuckets().stream().forEach(bucket -> {
+                bucket.getKey().forEach((k, v) -> {
+                    keys.get(k).add(v.toString());
```

Comment:
If the composite aggs request also includes the `*histograms`, then there is a strong possibility that the list of changed terms will have MANY duplicates.

Reply:
great find, actually 2 term group_by's introduced the same problem, I switched to sets to dedup on the fly